File: ios/Classes/FlutterBadgeControllerPlugin.swift
import Flutter
import UIKit

public class FlutterBadgeControllerPlugin: NSObject, FlutterPlugin {
  private static var badgeCount: Int = 0
  
  public static func register(with registrar: FlutterPluginRegistrar) {
    let channel = FlutterMethodChannel(name: "flutter_badge_controller", binaryMessenger: registrar.messenger())
    let instance = FlutterBadgeControllerPlugin()
    registrar.addMethodCallDelegate(instance, channel: channel)
  }

  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    switch call.method {
    case "setBadgeCount":
      if let args = call.arguments as? [String: Any],
        let count = args["count"] as? Int {
        FlutterBadgeControllerPlugin.badgeCount = count
        DispatchQueue.main.async {
          UIApplication.shared.applicationIconBadgeNumber = count
        }
      }
      result(nil)

    case "clearBadge":
      FlutterBadgeControllerPlugin.badgeCount = 0
      DispatchQueue.main.async {
        UIApplication.shared.applicationIconBadgeNumber = 0
      }
      result(nil)

    case "getBadgeCount":
      result(FlutterBadgeControllerPlugin.badgeCount)

    default:
      result(FlutterMethodNotImplemented)
    }
  }
}

File: android/src/main/kotlin/com/example/flutter_badge_controller/FlutterBadgeControllerPlugin.kt
package com.example.flutter_badge_controller

import android.app.NotificationManager
import android.content.Context
import androidx.annotation.NonNull
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result

/** FlutterBadgeControllerPlugin */
class FlutterBadgeControllerPlugin : FlutterPlugin, MethodChannel.MethodCallHandler {
  private lateinit var channel: MethodChannel
  private lateinit var context: Context
  private var badgeCount: Int = 0

  override fun onAttachedToEngine(@NonNull flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
    channel = MethodChannel(flutterPluginBinding.binaryMessenger, "flutter_badge_controller")
    channel.setMethodCallHandler(this)
    context = flutterPluginBinding.applicationContext
  }

  override fun onMethodCall(@NonNull call: MethodCall, @NonNull result: MethodChannel.Result) {
    when (call.method) {
      "setBadgeCount" -> {
        val count = call.argument<Int>("count") ?: 0
        setBadge(count)
        result.success(null)
      }
      "clearBadge" -> {
        setBadge(0)
        result.success(null)
      }
      "getBadgeCount" -> {
        result.success(badgeCount)
      }
      else -> {
        result.notImplemented()
      }
    }
  }

  private fun setBadge(count: Int) {
    badgeCount = count
    try {
      val launcherClassName = getLauncherClassName(context)
      if (launcherClassName != null) {
        val intent = android.content.Intent("android.intent.action.BADGE_COUNT_UPDATE")
        intent.putExtra("badge_count", count)
        intent.putExtra("badge_count_package_name", context.packageName)
        intent.putExtra("badge_count_class_name", launcherClassName)
        context.sendBroadcast(intent)
      }
    } catch (e: Exception) {
      e.printStackTrace()
    }
  }

  private fun getLauncherClassName(context: Context): String? {
    val pm = context.packageManager
    val intent = pm.getLaunchIntentForPackage(context.packageName) ?: return null
    return intent.component?.className
  }

  override fun onDetachedFromEngine(@NonNull binding: FlutterPlugin.FlutterPluginBinding) {
    channel.setMethodCallHandler(null)
  }
}
